packages/angular-cli/commands/completion.ts

```diff
@@ -1,17 +1,176 @@
+const lookupCommand = require('../ember-cli/lib/cli/lookup-command');
+
+function extractOptions(opts: any): String {
+  const output: String[] = [];
+
+  for (let index = 0; index < opts.length; index++) {
+    const element = opts[index];
+    output.push('--' + element.name);
+    if (element.aliases) {
+      output.push('-' + element.aliases[0]);
+    }
+  }
+
+  return output.sort().join(' ');
+}
+
+export interface CompletionCommandOptions {
+  all?: boolean;
+  bash?: boolean;
+  zsh?: boolean;
+};
+
+const commandsToIgnore = [
+  'easter-egg',
+  'destroy',
+  'github-pages-deploy' // errors because there is no base github-pages command
+];
+
+const optsNg: String[] = [];
 
 const CompletionCommand = Command.extend({
   name: 'completion',
   description: 'Adds autocomplete functionality to `ng` commands and subcommands',
   works: 'everywhere',
-  run: function() {
-    const scriptPath = path.resolve(__dirname, '..', 'utilities', 'completion.sh');
-    const scriptOutput = fs.readFileSync(scriptPath, 'utf8');
+  availableOptions: [
+    { name: 'all',   type: Boolean, default: true,  aliases: ['a'] },
+    { name: 'bash',  type: Boolean, default: false, aliases: ['b'] },
+    { name: 'zsh',   type: Boolean, default: false, aliases: ['z'] }
+  ],
+
+  run: function (commandOptions: CompletionCommandOptions) {
+    commandOptions.all = !commandOptions.bash && !commandOptions.zsh;
+
+    const commandFiles = fs.readdirSync(__dirname)
+      .filter(file => file.match(/\.ts$/) && !file.match(/\.run.ts$/))
+      .map(file => path.parse(file).name)
+      .filter(file => {
+        return commandsToIgnore.indexOf(file) < 0;
+      })
+      .map(file => file.toLowerCase());
+
+    const commandMap = commandFiles.reduce((acc: any, curr: string) => {
+      let classifiedName = stringUtils.classify(curr);
+      let defaultImport = require(`./${curr}`).default;
+
+      acc[classifiedName] = defaultImport;
+
+      return acc;
+    }, {});
+
+    let caseBlock = '';
+
+    commandFiles.forEach(cmd => {
+      const Command = lookupCommand(commandMap, cmd);
+      const com: String[] = [];
+
+      const command = new Command({
+        ui: this.ui,
+        project: this.project,
+        commands: this.commands,
+        tasks: this.tasks
+      });
+
+      optsNg.push(command.name);
+      com.push(command.name);
+
+      if (command.aliases) {
+        command.aliases.forEach((element: String) => {
+          optsNg.push(element);
+          com.push(element);
+        });
+      }
+
+      if (command.availableOptions && command.availableOptions[0]) {
+        let opts = extractOptions (command.availableOptions);
+        caseBlock = caseBlock + '    ' + com.sort().join('|') + ') opts="' + opts + '" ;;\n';
+      }
+    });
+
+    caseBlock = 'ng|help) opts="' + optsNg.sort().join(' ') + '" ;;\n' +
+                caseBlock +
+                '    *) opts="" ;;';
+
+    console.log(`###-begin-ng-completion###
```

Comment:
Could you use `stripIndent` from `common-tags` and indent the whole string? Look at other places where we use it for an example.